### PR TITLE
Add table that stores oracle script code information

### LINKED ISSUE #1036: Add table that stores oracle script code information
To prevent we load too much data to frontend. I propose we do the following.
- Create a `oracle_script_codes` table, with the following field: code_hash: bytes (pk), code_text: string, schema: json (or string? not sure). Both code_text and schema can be NULL and we will figure out how to update that later.
- For `oracle_scripts` table, let's store the hash of the code instead of the whole code. Make the column fk to the `oracle_script_codes` table.
- Whenver an oracle script is created or editted, we save the hash to the main table, but make sure to create a table for oracle_script_codes to have it ready for schema later. 

Also patch frontend for this

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed #1036

### CODE REVIEW

File: chain/db/db.go [hunk 1]
```diff
@@ -183,6 +191,13 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
+	db.Model(&RelatedDataSources{}).AddForeignKey(
```
Comment (taobun) [inline chain/db/db.go]: Already have this fk?

File: chain/db/table.go [hunk 2]
```diff
@@ -59,11 +60,11 @@ type DataSourceRevision struct {
 }
 
 type OracleScript struct {
-	ID          int64  `gorm:"primary_key;auto_increment:false"`
+	ID          int64  `gorm:"unique;auto_increment:false"`
```
Comment (taobun) [inline chain/db/table.go]: Please use this id as primary key

File: chain/db/table.go [hunk 3]
```diff
@@ -59,11 +60,11 @@ type DataSourceRevision struct {
...
 	Name        string `gorm:"not null"`
 	Description string `gorm:"not null"`
 	Owner       string `gorm:"not null"`
-	Code        []byte `gorm:"not null"`
+	CodeHash    []byte `gorm:"primary_key"`
```
Comment (taobun) [inline chain/db/table.go]: Just not null

File: chain/db/table.go (hunk 2 above)
Comment (sorawit) [inline chain/db/table.go]: Do not change this

**Review by sorawit — CHANGES_REQUESTED** [review]

File: chain/db/db.go [hunk 4]
```diff
@@ -78,9 +79,9 @@ func NewDB(dialect, path string, metadata map[string]string) (*BandDB, error) {
 		"RESTRICT",
 	)
 
-	db.Model(&OracleScriptRevision{}).AddForeignKey(
-		"tx_hash",
-		"transactions(tx_hash)",
+	db.Model(&OracleScript{}).AddForeignKey(
```
Comment (sorawit) [inline chain/db/db.go]: We should not remove OracleScriptRevision foreign key. Please double check your code before pushing for reviews.

**Review by sorawit — CHANGES_REQUESTED** [review]

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/oracle_script.go [hunk 5]
```diff
@@ -35,12 +39,30 @@ func (b *BandDB) AddOracleScript(
...
+	h := sha256.New()
+	h.Write(code)
+	codeHash := h.Sum(nil)
+	oracleScriptCode := OracleScriptCode{
+		CodeHash: codeHash,
+		CodeText: sql.NullString{},
+		Schema:   sql.NullString{},
+	}
+	var tmp OracleScriptCode
+	b.tx.First(&tmp, codeHash)
```
Comment (taobun) [inline chain/db/oracle_script.go]: FirstOrCreate?

**Review by taobun — CHANGES_REQUESTED** [review]

File: chain/db/oracle_script.go [hunk 6]
```diff
@@ -81,12 +101,33 @@ func (b *BandDB) handleMsgEditOracleScript(
 	msg zoracle.MsgEditOracleScript,
 	events map[string]string,
 ) error {
+	h := sha256.New()
+	h.Write(msg.Code)
+	codeHash := h.Sum(nil)
+
+	var oracleScriptCode OracleScriptCode
+	b.tx.First(&oracleScriptCode, codeHash)
```
Comment (taobun) [inline chain/db/oracle_script.go]: Use FirstOrCreate like create oracle script or declare helper function to do it

File: chain/db/oracle_script.go [hunk 7]
```diff
@@ -81,12 +101,33 @@ func (b *BandDB) handleMsgEditOracleScript(
...
+		msg.Owner, codeHash, b.ctx.BlockTime(),
 	)
 
-	err := b.tx.Save(&oracleScript).Error
+	err = b.tx.Save(&oracleScript).Error
+	if err != nil {
+		return err
+	}
+
+	err = b.tx.Save(&OracleScriptCode{
```
Comment (taobun) [inline chain/db/oracle_script.go]: Why save here? you just create this above?

**Review by taobun — CHANGES_REQUESTED** [review]

**Review by taobun — APPROVED** [review]

**Review by sorawit — APPROVED** [review]

**Pull request merged**